### PR TITLE
Fix ServiceEntry incremental updates, extend testing

### PR DESCRIPTION
**Pull request by howardjohn**
While implementing WorkloadEntry (https://github.com/istio/istio/pull/22531), I found a number of issues with ServiceEntry incremental updates when there are multiple ServiceEntries for the same host. Basically it would delete the other ServiceEntries and just keep the latest update until a full computation was triggered (not the same as a full push - this is only when a model.Service changes from a SE, its part of the SE implementation).

This PR fixes those issues. In the process, it adds a bunch of tests. Some of the changes to the test logic were made to support https://github.com/istio/istio/pull/22531, so they may not make complete sense in this PR (the TLS changes, mainly), but they don't hurt.

TODO:
* unit test updateExistingInstances. We have fully coverage probably, but nice to have simpler tests as well.
* make updateExistingInstances less horrible. This code is really ugly and inefficient right now
* make TestServiceDiscoveryServiceUpdate table driven instead of imperative, if possible

### CODE REVIEW

**Comment by howardjohn** [comment]
@ramaraochavali I got this working after many hours, but the code is pretty horrible.. if you have any suggestions let me know. It seems the main problem is we want to do a bunch of searches/deletions but we are doing this with arrays so we have a ton of reflect.DeepEquals in for loops everywhere

**Comment by howardjohn** [comment]
/retest

Curious if the same test fails

**Comment by ramaraochavali** [comment]
@howardjohn Sure. Will take a look and let you know.

**Comment by ramaraochavali** [comment]
@howardjohn Would it help if we change this map 
https://github.com/istio/istio/blob/master/pilot/pkg/serviceregistry/external/servicediscovery.go#L49 from 
`instances map[host.Name]map[string][]*model.ServiceInstance`
to 
`instances map[host.Name]map[string]map[string][]*model.ServiceInstance`
i.e. for each service entry keep track of instances added by it (by its name) - when a service entry is updated, we can do what we are doing today - delete instances added by it and add the new ones? I am thinking it might simplify that computation. WDYT?

**Comment by howardjohn** [comment]
does a service instance have a name or other unique idea identifier? if it
does that would make sense (although the type is really ugly, but it makes
sense)

**Comment by ramaraochavali** [comment]
ServiceInstanve does not have name. But I was thinking we should store with the name of the ServiceEntry that this service instances were generated from ( I am afk, do need to check if that model works to confirm though)
Agree, type looks really ugly- need to see if there is an alternative

**Comment by howardjohn** [comment]
/retest

infra flake

File: pilot/pkg/serviceregistry/external/servicediscovery.go [hunk 1]
```diff
@@ -37,6 +37,33 @@ var serviceEntryKind = collections.IstioNetworkingV1Alpha3Serviceentries.Resourc
 
 var _ serviceregistry.Instance = &ServiceEntryStore{}
 
+// instancesKey acts as a key to identify all instances for a given hostname/namespace pair
+// This is mostly used as an index
+type instancesKey struct {
```
Comment (ramaraochavali) [inline pilot/pkg/serviceregistry/external/servicediscovery.go:42]: nit: instancesKey -> instanceKey

File: pilot/pkg/serviceregistry/external/servicediscovery.go [hunk 2]
```diff
@@ -211,13 +244,15 @@ func (d *ServiceEntryStore) InstancesByPort(svc *model.Service, port int,
 
 	out := make([]*model.ServiceInstance, 0)
 
-	instances, found := d.instances[svc.Hostname][svc.Attributes.Namespace]
+	instanceLists, found := d.instances[instancesKey{svc.Hostname, svc.Attributes.Namespace}]
```
Comment (ramaraochavali) [inline pilot/pkg/serviceregistry/external/servicediscovery.go:247]: nit: use makeInstanceKey function here?

**Comment by tmshort** [comment]
Should this be backported to istio 1.5? We believe we might be running into this.

**Pull request merged**